### PR TITLE
Extended primitive core operations to support new types

### PR DESCRIPTION
**Pull request by abhishek-gola**
The support was already there, this PR tests them on edge cases and patch the fix.

closes: https://github.com/opencv/opencv/issues/24580
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/src/count_non_zero.dispatch.cpp [hunk 1]
```diff
@@ -198,7 +198,7 @@ void findNonZero(InputArray _src, OutputArray _idx)
         {
             const ushort* ptr16 = (const ushort*)ptr8;
             for( j = 0; j < cols; j++ )
-                if( (ptr16[j]<<1) != 0 ) buf[k++] = j;
```
Comment (asmorkalov) [inline modules/core/src/count_non_zero.dispatch.cpp:201]: Why? Please add a comment about the constant.

File: modules/core/test/test_arithm.cpp [hunk 2]
```diff
@@ -3585,6 +3585,84 @@ INSTANTIATE_TEST_CASE_P(
...
+typedef testing::TestWithParam<perf::MatDepth> NonZeroAccuracyNewTypes;
+
+TEST_P(NonZeroAccuracyNewTypes, accuracy)
+{
+    const int depth = GetParam();
+    const Size sz(123, 71);
+    cv::Mat src = cv::Mat::zeros(sz, CV_MAKETYPE(depth, 1));
+
+    std::vector<Point> expected_pts;
+    cv::RNG rng(0xCafeBabeULL);
```
Comment (asmorkalov) [inline modules/core/test/test_arithm.cpp]: you can use just theRNG. It's initialized by test system individually for each test case.

**Review by asmorkalov — APPROVED** [review]

**Review by asmorkalov — APPROVED** [review]
